### PR TITLE
[enhance](recycler) Refactor Accessor for Recycler

### PR DESCRIPTION
**Pull request by platoneko**
## Proposed changes

- Refactor `Accessor` for `Recycler`
- Fix missing initialized s3 accessor for storage vault in `InstanceRecycler`

### CODE REVIEW

File: cloud/src/recycler/storage_vault_accessor.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+struct FileMeta {
+    std::string path; // Relative path to accessor prefix
+    int64_t size {0};
+    int64_t mtime_s {0};
+};
+
+enum class AccessorType : uint8_t {
+    MOCK,
+    S3,
```
Comment (gavinchou) [inline cloud/src/recycler/storage_vault_accessor.h:39]: what about AZURE?

File: cloud/src/recycler/hdfs_accessor.cpp [hunk 2]
```diff
@@ -158,25 +159,145 @@ class HDFSBuilder {
...
+            level_dirs_.pop_back();
+        }
+
+        return res;
+    }
+
+private:
+    std::optional<DirEntries> list_directory(const char* dir_path) {
+        int num_entries;
+        auto* file_infos = hdfsListDirectory(hdfs_.get(), dir_path, &num_entries);
```
Comment (gavinchou) [inline cloud/src/recycler/hdfs_accessor.cpp:266]: When and where is `file_infos` reclaimed?

File: cloud/src/recycler/hdfs_accessor.cpp [hunk 3]
```diff
@@ -158,25 +159,145 @@ class HDFSBuilder {
...
+        }
+
+        auto& dir_entries = level_dirs_.back();
+        auto& entry = dir_entries.front();
+        DCHECK_EQ(entry.mKind, kObjectKindFile) << entry.mName;
+        std::string_view path {entry.mName};
+        DCHECK(path.starts_with(uri_)) << path << ' ' << uri_;
+        path.remove_prefix(uri_.length());
+        res = FileMeta {.path = std::string(path), .size = entry.mSize, .mtime_s = entry.mLastMod};
+        dir_entries.pop_front();
```
Comment (gavinchou) [inline cloud/src/recycler/hdfs_accessor.cpp:255]: Do we need to do some reclaiming work here?

File: cloud/src/recycler/checker.cpp [hunk 4]
```diff
@@ -358,36 +359,27 @@ int InstanceChecker::init(const InstanceInfoPB& instance) {
 
 int InstanceChecker::init_obj_store_accessors(const InstanceInfoPB& instance) {
     for (const auto& obj_info : instance.obj_info()) {
```
Comment (gavinchou) [inline cloud/src/recycler/checker.cpp:361]: log all obj_info here for better observability.
including instance id

File: cloud/src/recycler/checker.cpp [hunk 5]
```diff
@@ -396,52 +388,56 @@ int InstanceChecker::init_storage_vault_accessors(const InstanceInfoPB& instance
         return 0;
     }
 
-    std::string storage_vault_start = storage_vault_key({instance_id_, ""});
-    std::string storage_vault_end = storage_vault_key({instance_id_, "\xff"});
-    std::unique_ptr<RangeGetIterator> it;
+    FullRangeGetIteratorOptions opts(txn_kv_);
```
Comment (gavinchou) [inline cloud/src/recycler/checker.cpp:391]: is `full_range_get()` tested well enough?

**Comment by gavinchou** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Comment by ByteYue** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by ByteYue — APPROVED** [review]

**Pull request merged**